### DP

동적 계획법(Dynamic Programming)은 복잡한 문제를 해결하기 위해 사용되는 알고리즘 디자인 기법입니다.

동적 계획법은 하위 문제(subproblem)의 최적해를 계산하고, 그 결과를 이용하여 큰 문제의 최적해를 구하는 방식으로 작동합니다.

이를 통해 중복되는 하위 문제들을 효율적으로 해결하고, 전체 문제의 답을 구할 수 있습니다.



동적 계획법은 다음과 같은 상황에서 유용하게 사용될 수 있습니다:

중복되는 하위 문제: 문제를 작은 하위 문제로 분할했을 때, 동일한 하위 문제들이 반복적으로 해결되는 경우 동적 계획법을 적용할 수 있습니다.

최적 부분 구조: 큰 문제의 최적해가 작은 문제의 최적해들로부터 구성될 수 있는 경우 동적 계획법을 사용할 수 있습니다.

동적 계획법은 일반적으로 다음과 같은 단계를 따릅니다:

문제를 작은 하위 문제들로 분할합니다.

하위 문제의 최적해를 계산하기 위해 재귀적인 방식이나 반복적인 방식을 사용합니다.

작은 하위 문제들의 최적해를 이용하여 큰 문제의 최적해를 구합니다.

파이썬으로 동적 계획법을 구현하는 방법을 살펴보겠습니다.




In [1]:
# 동적 계획법을 사용한 피보나치 수열 계산

def fibonacci(n):
    # n번째 피보나치 수를 저장하기 위한 배열
    fib = [0] * (n+1)

    # 초기값 설정
    fib[0] = 0
    fib[1] = 1

    # 작은 하위 문제들의 최적해를 이용하여 큰 문제의 최적해를 구함
    for i in range(2, n+1):
        fib[i] = fib[i-1] + fib[i-2]

    return fib[n]

# 예시 실행
n = 10
result = fibonacci(n)
print("피보나치 수열의", n, "번째 수:", result)


피보나치 수열의 10 번째 수: 55


위의 예시에서는 fib 배열을 사용하여 하위 문제의 최적해를 저장합니다.

반복문을 통해 작은 문제들의 최적해를 계산하고, 이를 이용하여 큰 문제의 최적해를 구합니다.

이러한 방식으로 동적 계획법을 활용하여 피보나치 수열의 n번째 수를 효율적으로 계산할 수 있습니다.






### 알고리즘 개념을 숲을 보는 시점으로 생각하기.


알고리즘 개념을 숲을 보는 시점으로 생각하는 것은 문제 해결 과정을 전체적인 큰 그림으로 이해하는 것을 의미합니다.

숲을 보는 시점에서는 알고리즘의 목적과 작동 방식, 그리고 문제 해결에 필요한 전체적인 단계를 이해할 수 있습니다.

일반적으로 알고리즘은 작은 단계들의 연속이며, 각 단계는 입력을 처리하여 출력을 생성하는 과정을 나타냅니다.

숲을 보는 시점에서는 다음과 같은 관점을 갖을 수 있습니다:

1. 목적: 알고리즘은 특정한 문제를 해결하기 위해 설계됩니다.

  숲을 보는 시점에서는 알고리즘의 목적이 무엇인지 이해할 수 있습니다.

  이를 통해 문제 해결에 필요한 전반적인 방향성을 파악할 수 있습니다.

2. 구성 요소: 알고리즘은 작은 구성 요소들의 집합으로 이루어집니다.

  이러한 구성 요소들은 문제를 해결하기 위해 필요한 단계들을 나타내며, 서로 상호작용하면서 전체적인 문제 해결 과정을 형성합니다.

  숲을 보는 시점에서는 알고리즘의 구성 요소들을 파악하여 전체적인 흐름을 이해할 수 있습니다.

3. 작동 방식: 알고리즘은 입력을 받아 처리하고, 특정한 규칙이나 절차에 따라 연산을 수행하여 원하는 결과를 도출합니다.

  숲을 보는 시점에서는 알고리즘의 작동 방식이 어떤지 이해할 수 있으며, 이를 통해 문제를 해결하는 과정을 이해할 수 있습니다.

4. 복잡도: 알고리즘의 복잡도는 알고리즘의 효율성과 관련이 있습니다.

  숲을 보는 시점에서는 알고리즘의 복잡도를 고려하여 어떤 입력에 대해서도 효율적으로 동작할 수 있는지 판단할 수 있습니다.

알고리즘 개념을 숲을 보는 시점으로 바라보는 것은 문제 해결에 필요한 전체적인 그림을 이해하는 데 도움이 됩니다.

이를 통해 알고리즘의 목적과 작동 방식을 파악하고, 문제를 해결하기 위한 전체적인 단계를 이해할 수 있습니다.

### DP와 Greedy 사례

동적 계획법 사례: 시계열 데이터 예측 (ARIMA 모델)

ARIMA(AutoRegressive Integrated Moving Average) 모델은 동적 계획법을 활용하여 시계열 데이터 예측에 사용되는 일반적인 모델입니다.

ARIMA 모델은 자기회귀(AR, Autoregressive)와 이동평균(MA, Moving Average)의 조합으로 구성됩니다.

아래는 ARIMA 모델을 사용하여 시계열 데이터를 예측하는 예시 코드입니다.



In [2]:
import pandas as pd
from statsmodels.tsa.arima.model import ARIMA

# 예시 데이터 생성
data = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]

# ARIMA 모델 훈련
model = ARIMA(data, order=(1, 0, 0))  # (AR, 차분, MA) 순서
model_fit = model.fit()

# 다음 값 예측
prediction = model_fit.predict(len(data), len(data))
print("다음 값 예측:", prediction[0])


다음 값 예측: 98.70025578567268


위의 예시 코드는 statsmodels 패키지의 ARIMA 모델을 사용하여 시계열 데이터의 다음 값을 예측하는 방법을 보여줍니다.



그리디 알고리즘 사례: 특징 선택 (Forward Selection)

특징 선택은 데이터 분석에서 중요한 변수를 선택하고 불필요한 변수를 제거하여 모델의 복잡성을 줄이는 방법입니다.

그리디 알고리즘 중 Forward Selection 방법은 변수를 한 번에 하나씩 추가하여 최적의 특징 집합을 찾아갑니다.

아래는 Forward Selection을 사용하여 특징 선택을 수행하는 예시 코드입니다.



In [3]:
import numpy as np
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression

# 예시 데이터 로드
iris = load_iris()
X = iris.data
y = iris.target

# Forward Selection 알고리즘
selected_features = []
num_features = X.shape[1]

for _ in range(num_features):
    best_accuracy = 0
    best_feature = None

    for feature in range(num_features):
        if feature not in selected_features:
            features = selected_features + [feature]
            X_selected = X[:, features]
            model = LogisticRegression()
            model.fit(X_selected, y)
            accuracy = model.score(X_selected, y)

            if accuracy > best_accuracy:
                best_accuracy = accuracy
                best_feature = feature

    selected_features.append(best_feature)

print("선택된 특징 인덱스:", selected_features)


선택된 특징 인덱스: [3, 2, 1, 0]


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


위의 예시 코드는 붓꽃(iris) 데이터셋에서 Forward Selection을 사용하여 특징 선택을 수행하는 방법을 보여줍니다.

Logistic Regression을 사용하여 각 특징을 추가하여 최적의 특징 집합을 선택합니다.

위의 코드는 동적 계획법과 그리디 알고리즘이 데이터 분석에서 활용되는 사례와 간단한 파이썬 코드 예시를 제시하였습니다.

다양한 데이터 분석 문제에 적용할 수 있는 다양한 알고리즘과 기법들이 있으며, 문제의 특성에 따라 적합한 알고리즘을 선택하여 사용할 수 있습니다.






동적 계획법 사례: 최장 증가 부분 수열 (Longest Increasing Subsequence)

최장 증가 부분 수열(LIS)은 동적 계획법을 활용하여 구할 수 있는 대표적인 문제입니다.

LIS는 주어진 수열에서 증가하는 순서로 구성된 가장 긴 부분 수열을 찾는 문제입니다.

아래는 동적 계획법을 사용하여 최장 증가 부분 수열을 구하는 예시 코드입니다.



In [4]:
def longest_increasing_subsequence(nums):
    n = len(nums)
    dp = [1] * n  # 각 인덱스에서 끝나는 LIS의 길이를 저장하는 배열

    for i in range(1, n):
        for j in range(i):
            if nums[i] > nums[j]:
                dp[i] = max(dp[i], dp[j] + 1)

    return max(dp)

# 예시 실행
sequence = [3, 1, 4, 1, 5, 9, 2, 6, 5]
result = longest_increasing_subsequence(sequence)
print("최장 증가 부분 수열의 길이:", result)


최장 증가 부분 수열의 길이: 4


위의 코드는 주어진 수열에서 최장 증가 부분 수열의 길이를 구하는 동적 계획법 기반의 알고리즘입니다.

주어진 수열을 순회하면서 각 숫자에 대해 이전 숫자들과 비교하여 LIS의 길이를 갱신합니다.



그리디 알고리즘 사례: 배낭 문제 (Knapsack Problem)

배낭 문제는 그리디 알고리즘을 활용하여 해결할 수 있는 대표적인 문제입니다.

배낭 문제는 주어진 가치와 무게를 갖는 물건들을 배낭에 넣을 때, 가치의 합이 최대가 되도록 물건을 선택하는 문제입니다.

아래는 그리디 알고리즘을 사용하여 배낭 문제를 해결하는 예시 코드입니다.



In [5]:
def knapsack(capacity, values, weights):
    n = len(values)
    items = list(zip(values, weights))
    items.sort(key=lambda x: x[0] / x[1], reverse=True)  # 가치 단위 무게 비율을 기준으로 정렬

    max_value = 0
    selected_items = []

    for value, weight in items:
        if capacity >= weight:
            max_value += value
            capacity -= weight
            selected_items.append((value, weight))

    return max_value, selected_items

# 예시 실행
capacity = 10
values = [40, 100, 50, 60]
weights = [2, 5, 10, 6]
result_value, result_items = knapsack(capacity, values, weights)
print("최대 가치:", result_value)
print("선택된 물건:", result_items)


최대 가치: 140
선택된 물건: [(40, 2), (100, 5)]


위의 코드는 주어진 배낭 용량과 물건들의 가치 및 무게를 기반으로 그리디 알고리즘을 사용하여 배낭 문제를 해결합니다.

가치 단위 무게 비율을 기준으로 물건들을 정렬하고, 순서대로 배낭에 넣을 수 있는지 확인하면서 최대 가치를 구합니다.



### 핸즈온

In [16]:
# 메모이제이션 코드
def memo_fib(input_value, save_memo):

    if input_value in save_memo:
        return save_memo[input_value]
    if input_value == 1 or input_value == 2:
        value = 1
    elif input_value > 2:
        value = memo_fib(input_value - 1, save_memo) + memo_fib(input_value - 2, save_memo)
    save_memo[input_value] = value
    return value

# 태뷸레이션 코드
def tabul_fib(input_value):
    f = [0, 1]
    for i in range(2, input_value + 1):
        f.append(f[i - 1] + f[i - 2])
    return f[input_value]


In [17]:
save_memo = {}  # 메모이제이션을 위한 빈 딕셔너리 생성

result_memo = memo_fib(33, save_memo)
print("메모이제이션 결과:", result_memo)

result_tabul = tabul_fib(33)
print("태뷸레이션 결과:", result_tabul)


메모이제이션 결과: 3524578
태뷸레이션 결과: 3524578


memo_fib 함수는 재귀적인 방식으로 계산하면서 중복 계산을 피하고, tabul_fib 함수는 반복문을 통해 이전 값들을 참조하여 계산합니다.






### 예시

예를 들어, 계단을 오를 때 한 번에 한 계단 또는 두 계단씩 오를 수 있다고 가정해봅시다.

계단의 개수가 주어졌을 때, 최소한의 계단 오르기 횟수로 꼭대기까지 도달하는 방법을 찾는 문제를 DP로 해결할 수 있습니다.



In [1]:
def min_stair_steps(n):
    if n == 1:  # 계단이 한 개인 경우
        return 1
    if n == 2:  # 계단이 두 개인 경우
        return 2

    # DP 테이블 초기화
    dp = [0] * (n + 1)
    dp[1] = 1
    dp[2] = 2

    # DP 테이블 채우기
    for i in range(3, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]

    return dp[n]

n = 5
result = min_stair_steps(n)
print(f"계단 오르는 경우의 수: {result}")


계단 오르는 경우의 수: 8


위 코드에서 min_stair_steps 함수는 DP를 사용하여 계단 오르기 횟수를 구하는 함수입니다.

DP 테이블을 활용하여 작은 계단부터 큰 계단까지 순차적으로 계산합니다.



    1 -> 2 -> 3 -> 4 -> 5 : 1계단씩오르기

    1 -> 2 -> 3 -> 5 : 1계단씩오르다가 마지막에 2계단

    1 -> 2 -> 4 -> 5 : 1계단씩오르다가 2계단 1계단

    1 -> 3 -> 4 -> 5

    1 -> 3 -> 5 : 2계단 2계단

    2 -> 3 -> 4 -> 5

    2 -> 3 -> 5

    2 -> 4 -> 5

그리디 알고리즘은 각 단계에서 가장 좋은 선택을 하는 방식으로 문제를 해결합니다.

각 선택이 지역적으로는 최적이지만 전역적으로는 최적해를 보장하지 않을 수 있습니다.


예를 들어, 동전을 사용하여 특정 금액을 만드는 문제를 그리디로 해결해봅시다.

이때, 동전은 500원, 100원, 50원, 10원으로 가정합니다.



In [20]:
def min_coin_count(target):
    coin_types = [500, 100, 50, 10]  # 동전의 종류

    total_count = 0
    for coin in coin_types:
        count = target // coin  # 해당 동전으로 거슬러 줄 수 있는 개수
        total_count += count
        target %= coin  # 남은 금액

    return total_count

target = 1260
result = min_coin_count(target)
print(f"동전의 최소 개수: {result}")


동전의 최소 개수: 6


위 코드에서 min_coin_count 함수는 그리디 알고리즘을 사용하여 동전의 최소 개수를 구하는 함수입니다.

가장 큰 동전부터 사용하여 거슬러 줄 수 있는 개수를 계산하고, 남은 금액으로 반복하여 최소 개수를 구합니다.



그리디 알고리즘을 사용한 또 다른 예시로 "최대 수입 스케줄링 문제 (Maximum Revenue Scheduling Problem)"를 살펴보겠습니다.

주어진 일정과 해당 일정에서 얻을 수 있는 수입을 고려하여 최대한 많은 수입을 얻는 문제입니다.



In [30]:
def max_revenue_schedule(jobs):
    jobs.sort(key=lambda x: x[1], reverse=True)  # 수입을 기준으로 내림차순 정렬

    max_revenue = 0
    selected_jobs = []

    for job in jobs:
        if job[0] >= len(selected_jobs):  # 현재 일정이 선택된 일정과 겹치지 않는 경우
            max_revenue += job[1]
            selected_jobs.append(job)

    return max_revenue

jobs = [(3, 100), (1, 50), (2, 120), (4, 70), (2, 80)]

result = max_revenue_schedule(jobs)
print(f"최대 수입: {result}")



최대 수입: 370


위 코드에서 max_revenue_schedule 함수는 그리디 알고리즘을 사용하여 최대 수입을 얻는 함수입니다.

일정을 수입을 기준으로 내림차순으로 정렬한 후, 선택된 일정과 겹치지 않는 일정을 선택하며 최대 수입을 누적합니다.



주어진 일정과 수입 리스트: [(3, 100), (1, 50), (2, 120), (4, 70), (2, 80)]
최대 수입: 0
선택된 일정: []

수입을 기준으로 내림차순으로 정렬합니다.

(2, 120)을 선택:
- 현재 선택한 일정: [(2, 120)]
- 선택된 일정의 길이(1) >= 현재 일정의 시작일(2)인 경우 선택 가능
  - 최대 수입: 0 + 120 = 120

(3, 100)을 선택:
- 현재 선택한 일정: [(2, 120), (3, 100)]
- 선택된 일정의 길이(2) >= 현재 일정의 시작일(3)인 경우 선택 가능
  - 최대 수입: 120 + 100 = 220

(2, 80)을 선택:
- 현재 선택한 일정: [(2, 120), (3, 100), (2, 80)]
- 선택된 일정의 길이(3) >= 현재 일정의 시작일(2)인 경우 선택 가능
  - 최대 수입: 220 + 80 = 300

(4, 70)을 선택:
- 현재 선택한 일정: [(2, 120), (3, 100), (2, 80), (4, 70)]
- 선택된 일정의 길이(4) >= 현재 일정의 시작일(4)인 경우 선택 가능
  - 최대 수입: 300 + 70 = 370

최대 수입: 370

